### PR TITLE
Add arm64/aarch64 support

### PR DESCRIPTION
**Pull request by abiosoft**
Fixes https://github.com/lima-vm/alpine-lima/issues/26.

This introduces an additional `ARCH` value that defaults to the output of `uname -m`. e.g. `x86_64`, `aarch64`.

~I contemplated accepting `aarch64` as value instead  of `arm64` but it requires more code changes. I opted for the easier way out.~

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,8 +1,13 @@
 ARG ALPINE_VERSION=latest
 FROM alpine:${ALPINE_VERSION}
 RUN apk add alpine-sdk build-base apk-tools alpine-conf busybox \
-  fakeroot syslinux xorriso squashfs-tools sudo \
+  fakeroot xorriso squashfs-tools sudo \
   mtools dosfstools grub-efi
+
+# syslinux is missing for aarch64
+ARG ARCH=amd64
+RUN if [ "${ARCH}" = "amd64" ]; then apk add syslinux; fi
```
Comment (AkihiroSuda) [inline Dockerfile:9]: Can we use predefined `$TARGETARCH`?
https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope

File: Makefile [hunk 2]
```diff
@@ -8,22 +8,27 @@ BUILD_ID ?= $(shell git describe --tags)
 # len("alpine-lima-12345-3.13.5-x86_64") == 31
 EDITION ?= std
 
+ARCH ?= amd64
```
Comment (AkihiroSuda) [inline Makefile:11]: Can we detect the default ARCH from `uname -m`?

File: Makefile [hunk 3]
```diff
@@ -8,22 +8,28 @@ BUILD_ID ?= $(shell git describe --tags)
...
 NERDCTL_VERSION=0.12.1
 
 .PHONY: mkimage
 mkimage:
 	cd src/aports && git fetch && git checkout $(GIT_TAG)
 	docker build \
-		--tag mkimage:$(ALPINE_VERSION) \
+		--tag mkimage:$(ALPINE_VERSION)-$(ARCH) \
 		--build-arg ALPINE_VERSION=$(ALPINE_VERSION) \
+		--platform linux/$(ARCH) \
```
Comment (AkihiroSuda) [inline Makefile:23]: The `--platform` value has to be like `linux/amd64`, so `uname -m` value such as `x86_64` can't be passed here

Comment (AkihiroSuda) [reply Makefile:23]: Can we test `make`-ing the iso using GitHub Actions, at least for `amd64`?

Comment (abiosoft) [reply Makefile:23]: > The --platform value has to be like linux/amd64, so uname -m value such as x86_64 can't be passed here

docker actually support `linux/x86_64`, linux/amd64`, `linux/aarch64` and `linux/arm64` from my tests. But I will make it consistent just to be on the safe side.

> Can we test make-ing the iso using GitHub Actions, at least for amd64?

I am not sure how to trigger this

Comment (jandubois) [reply Makefile]: > Can we test `make`-ing the iso using GitHub Actions, at least for `amd64`?

This should be a separate PR (see #10).

**Comment by jandubois** [comment]
Are you sure the `aarch64` image actually works? `syslinux` is the boot loader used by Alpine, and without it the ISO doesn't load for me (using Lima):

```
Boot failed: Could not read from CDROM (code 0004) Booting from Hard Disk...
Boot failed: not a bootable disk
Booting from ROM...
iPXE (PCI 00:01.0) starting execution...ok iPXE initialising devices...ok
iPXE 1.20.1+ Cg4bd0) — Open Source Network Boot Firmware — http://ipxe.org Features: DNS HTTP iSCSI TFTP AoE ELF MBOOT PXE bzlmage Menu PXEXT
net0: 52:55:55:66:d7:cd using virtio-net on 0000:00:01.0 (open) [Link :up, TX:0 TXE:0 RX:0 RXE:01
Configuring (net0 52 :55 :55 :66 :d?:cd)..... ok
net0: 192.160.5.15/255.255.255.0 gw 192.160.5.2
Nothing to boot: No such file or directory (http://ipxe.org/2d03el3b) No more network devices
Booting from Floppy...
Boot failed: could not read the boot disk
No bootable device.
```

**Comment by abiosoft** [comment]
> Are you sure the `aarch64` image actually works? `syslinux` is the boot loader used by Alpine, and without it the ISO doesn't load for me (using Lima):

My bad, I did not try out the `lima` and `run` targets but I successfully used the generated iso with Lima. 
I will verify that.

**Comment by AkihiroSuda** [comment]
> Are you sure the `aarch64` image actually works? `syslinux` is the boot loader used by Alpine, and without it the ISO doesn't load for me (using Lima):
> 
> ```
> Boot failed: Could not read from CDROM (code 0004) Booting from Hard Disk...
> Boot failed: not a bootable disk
> Booting from ROM...
> iPXE (PCI 00:01.0) starting execution...ok iPXE initialising devices...ok
> iPXE 1.20.1+ Cg4bd0) — Open Source Network Boot Firmware — http://ipxe.org Features: DNS HTTP iSCSI TFTP AoE ELF MBOOT PXE bzlmage Menu PXEXT
> net0: 52:55:55:66:d7:cd using virtio-net on 0000:00:01.0 (open) [Link :up, TX:0 TXE:0 RX:0 RXE:01
> Configuring (net0 52 :55 :55 :66 :d?:cd)..... ok
> net0: 192.160.5.15/255.255.255.0 gw 192.160.5.2
> Nothing to boot: No such file or directory (http://ipxe.org/2d03el3b) No more network devices
> Booting from Floppy...
> Boot failed: could not read the boot disk
> No bootable device.
> ```

Your boot log seems qemu-system-x86\_64 ,  not qemu-system-aarch64

**Comment by jandubois** [comment]
> Your boot log seems qemu-system-x86_64 , not qemu-system-aarch64

Ah yes, thanks for pointing that out (I've just being trying to get things to work using `u-boot` instead of `syslinux`, not realizing my mistake).

The `lima.sh` script needs to specify the arch as well; let me try again.

**Comment by jandubois** [comment]
> successfully used the generated iso with Lima

Indeed, it does work with lima, after some more fixes to `lima` and `iso` targets:

```diff
diff --git Makefile Makefile
index 2e29667..da7de73 100644
--- Makefile
+++ Makefile
@@ -38,13 +38,13 @@ nerdctl-$(NERDCTL_VERSION)-$(ARCH):

 .PHONY: lima
 lima:
-       ALPINE_VERSION=$(ALPINE_VERSION) EDITION=$(EDITION) ./lima.sh
+       ALPINE_VERSION=$(ALPINE_VERSION) EDITION=$(EDITION) ARCH=$(ARCH) ./lima.sh

 .PHONY: run
 run:
-       qemu-system-x86_64 \
+       qemu-system-$(ARCH_ALIAS) \
                -boot order=d,splash-time=0,menu=on \
-               -cdrom iso/alpine-lima-$(EDITION)-$(ALPINE_VERSION)-x86_64.iso \
+               -cdrom iso/alpine-lima-$(EDITION)-$(ALPINE_VERSION)-$(ARCH).iso \
                -cpu host \
                -machine q35,accel=hvf \
                -smp 4,sockets=1,cores=4,threads=1 \
diff --git lima.sh lima.sh
index 4d5cdc4..2ce85d2 100755
--- lima.sh
+++ lima.sh
@@ -2,9 +2,10 @@
 set -eu

 cat <<EOF >"${EDITION}.yaml"
+arch: "${ARCH}"
 images:
-- location: "${PWD}/iso/alpine-lima-${EDITION}-${ALPINE_VERSION}-x86_64.iso"
-  arch: "x86_64"
+- location: "${PWD}/iso/alpine-lima-${EDITION}-${ALPINE_VERSION}-${ARCH}.iso"
+  arch: "${ARCH}"
 mounts:
 - location: "~"
   writable: false
```

I do not get a working console with it, but that is really just a debugging thing when the image doesn't work with Lima, so maybe we can ignore this for now.

**Review by jandubois — APPROVED** [review]
Thanks, LGTM!

@AkihiroSuda do you have any more feedback, or should we merge now?

**Comment by AkihiroSuda** [comment]
Looks good, but could you squash commits?
Thanks

**Comment by abiosoft** [comment]
> Looks good, but could you squash commits? Thanks

Done

**Review by AkihiroSuda — APPROVED** [review]
Thanks
